crates/rethnet/tests/cli.rs

```diff
@@ -0,0 +1,179 @@
+
+    // the order of operations is a little weird in this test, because we spawn a separate process
+    // for the server, and we want to make sure that we end that process gracefully. more
+    // specifically, once the server is started, we avoid the ? operator until the server has been
+    // stopped.
+
+    // hold method invocations separately from requests so that we can easily iterate over them in
+    // order to check for corresponding log entries in the server output:
+    let method_invocations = [
+        MethodInvocation::Eth(EthMethodInvocation::GetBalance(
+            address,
+            BlockSpec::Tag(String::from("latest")),
+        )),
+        MethodInvocation::Eth(EthMethodInvocation::GetCode(
+            address,
+            BlockSpec::Tag(String::from("latest")),
+        )),
+        MethodInvocation::Eth(EthMethodInvocation::GetStorageAt(
+            address,
+            U256::ZERO,
+            BlockSpec::Tag(String::from("latest")),
+        )),
+        MethodInvocation::Eth(EthMethodInvocation::GetTransactionCount(
+            address,
+            BlockSpec::Tag(String::from("latest")),
+        )),
+        MethodInvocation::Hardhat(HardhatMethodInvocation::SetBalance(address, U256::ZERO)),
+        MethodInvocation::Hardhat(HardhatMethodInvocation::SetCode(
+            address,
+            Bytes::from_static("deadbeef".as_bytes()).into(),
+        )),
+        MethodInvocation::Hardhat(HardhatMethodInvocation::SetNonce(address, U256::ZERO)),
+        MethodInvocation::Hardhat(HardhatMethodInvocation::SetStorageAt(
+            address,
+            U256::ZERO,
+            U256::ZERO,
+        )),
+    ];
+
+    // prepare request body before even spawning the server because serialization could fail:
+    let request_body: String = serde_json::to_string(
+        &method_invocations
+            .iter()
+            .enumerate()
+            .map(|(id, method)| RpcRequest {
+                version: jsonrpc::Version::V2_0,
+                id: jsonrpc::Id::Num(id.try_into().unwrap()),
+                method: method.clone(),
+            })
+            .collect::<Vec<RpcRequest<MethodInvocation>>>(),
+    )?;
+
+    // spawn the server process:
+    let mut server = Command::cargo_bin("rethnet")?
+        .arg("node")
+        .stdout(Stdio::piped())
+        .spawn()?;
+
+    // (required for CI runs on MacOS) sleep a moment to make sure the server comes up before we
+    // start sending requests:
+    std::thread::sleep(std::time::Duration::from_secs(1));
+
+    // query the server, but don't check the Result yet, because returning early would prevent us
+    // from gracefully terminating the server:
+    let send_result = reqwest::Client::new()
+        .post("http://127.0.0.1:8545/")
+        .header(reqwest::header::CONTENT_TYPE, "application/json")
+        .body(request_body)
+        .send()
+        .await;
+
+    // signal the server to shut down gracefully:
+    shutdown(&mut server)?;
```

Comment:
If this is a repeating pattern for tests, you can create a struct that implements the `Drop` trait and calls  shutdown` on drop. It makes it more explicit and doesn't require the comments to explain what's going on, aka self-explanatory code.

Reply:
the thing is that we need to use that `server` object even after the shutdown (see the next line) in order to get the standard output of the process, so we can validate its log entries, so i think shutting down the process via `Drop` doesn't fit what we need here. or do you see some other way to utilize `Drop`?